### ISSUE TITLE
Issue: Couldn't make JDWP connection with Vulkan on Android (7.0)

### ISSUE THREAD

**Issue by Nielsbishere** [baldurk/renderdoc#996]
## Description

When I run my Vulkan Android app (with debug layers), it runs fine and on PC it does as well. On PC, I can use RenderDoc as intended, but when I launch Renderdoc and debug my application it crashes. If I forcefully run my program and check my log function, it tells me it can't find a Vulkan layer. I think it is referring to libVkLayer_GLES_RenderDoc.so, as it mentions it in the RenderDoc log.
```
RDOC 004356: [19:04:53]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward tcp:39970 localabstract:renderdoc_39920'
RDOC 004356: [19:04:53]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward tcp:38970 localabstract:renderdoc_38920'
RDOC 004356: [19:04:55]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell setprop debug.vulkan.layers :'
RDOC 004356: [19:05:00]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward tcp:39970 localabstract:renderdoc_39920'
RDOC 004356: [19:05:00]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward tcp:38970 localabstract:renderdoc_38920'
RDOC 004356: [19:05:08]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward tcp:39970 localabstract:renderdoc_39920'
RDOC 004356: [19:05:08]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward tcp:38970 localabstract:renderdoc_38920'
RDOC 004356: [19:05:11]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell pm list packages -3'
RDOC 004356: [19:05:14]    android_patch.cpp( 497) - Log     - Checking that APK is debuggable
RDOC 004356: [19:05:14]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell dumpsys package net.osomi.Osomi_Core'
RDOC 004356: [19:05:15]    android_patch.cpp( 460) - Log     - Checking for root access on ce051605ada61d2805
RDOC 004356: [19:05:15]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 root'
RDOC 004356: [19:05:15]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell whoami'
RDOC 004356: [19:05:15]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell test -e /system/xbin/su && echo found'
RDOC 004356: [19:05:15]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward tcp:39970 localabstract:renderdoc_39920'
RDOC 004356: [19:05:15]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward tcp:38970 localabstract:renderdoc_38920'
RDOC 004356: [19:05:19]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell cmd package resolve-activity -c android.intent.category.LAUNCHER net.osomi.Osomi_Core'
RDOC 004356: [19:05:19]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward --remove tcp:39024'
RDOC 004356: [19:05:19]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell am force-stop net.osomi.Osomi_Core'
RDOC 004356: [19:05:20]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell setprop debug.vulkan.layers VK_LAYER_RENDERDOC_Capture'
RDOC 004356: [19:05:20]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell mkdir -p /sdcard/Android/data/net.osomi.Osomi_Core'
RDOC 004356: [19:05:20]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell setprop debug.rdoc.RENDERDOC_CAPTUREOPTS abababababaaaaaaaaaaaaaaababababababcnip'
RDOC 004356: [19:05:20]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell pm path net.osomi.Osomi_Core'
RDOC 004356: [19:05:21]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell ls /data/app/net.osomi.Osomi_Core-2//lib/*/libVkLayer_GLES_RenderDoc.so'
RDOC 004356: [19:05:21]          android.cpp( 173) - Log     - No library found in /data/app/net.osomi.Osomi_Core-2//lib/*/libVkLayer_GLES_RenderDoc.so for net.osomi.Osomi_Core - assuming injection is required.
RDOC 004356: [19:05:21]          android.cpp( 185) - Log     - Setting up to launch the application as a debugger to inject.
RDOC 004356: [19:05:21]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell am start -S -D net.osomi.Osomi_Core/android.app.NativeActivity'
RDOC 004356: [19:05:22]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell ps -A | grep net.osomi.Osomi_Core'
RDOC 004356: [19:05:22]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell ps -A | grep net.osomi.Osomi_Core'
RDOC 004356: [19:05:23]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell ps -A | grep net.osomi.Osomi_Core'
RDOC 004356: [19:05:23]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell ps -A | grep net.osomi.Osomi_Core'
RDOC 004356: [19:05:23]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward tcp:39970 localabstract:renderdoc_39920'
RDOC 004356: [19:05:23]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward tcp:38970 localabstract:renderdoc_38920'
RDOC 004356: [19:05:23]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 shell ps -A | grep net.osomi.Osomi_Core'
RDOC 004356: [19:05:23]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward tcp:39970 localabstract:renderdoc_39920'
RDOC 004356: [19:05:23]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward tcp:38970 localabstract:renderdoc_38920'
RDOC 004356: [19:05:24]             jdwp.cpp( 399) - Error   - Couldn't make JDWP connection
RDOC 004356: [19:05:24]          android.cpp( 217) - Error   - Failed to inject using JDWP
RDOC 004356: [19:05:30]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward tcp:39970 localabstract:renderdoc_39920'
RDOC 004356: [19:05:30]    android_tools.cpp( 283) - Log     - COMMAND: D:/programs/Android/sdk/platform-tools/adb '-s ce051605ada61d2805 forward tcp:38970 localabstract:renderdoc_38920'
```
`No library found in /data/app/net.osomi.Osomi_Core-2//lib/*/libVkLayer_GLES_RenderDoc.so for net.osomi.Osomi_Core - assuming injection is required.` seems weird to me, as I am not even using GLES, I am using Vulkan.

This keeps on happening if I open my app, even if RenderDoc is only on in the background, but when I close it, it just runs normally.

## Repro steps

Simply going to [RenderDoc (1.0)](https://github.com/baldurk/renderdoc/commit/5ef2d0bb5e86a37ba6c250987e9dc8c33c26435c) and selecting my device in the dropdown. Then launching it and selecting and running my application from the executable path. 

## Environment

RenderDoc build: v1.0
Operating system: Windows 8.1 x64
Target system: Android 7.0 (Level 24) armv8a
API: Vulkan 1.0
GPU: Mali-T880 (Samsung S7 Edge)

**Comment by baldurk** [comment]
I believe your core issue is that your device doesn't support `ps -A`, and the target PID was not found so the debug connection couldn't be made. This was fixed a while ago after v1.0 was released, so trying any recent nightly build should eliminate it as a possible problem - there are also some other fixes to android that have been made since v1.0.

Note though that I believe #993 introduced a 100% crash bug that has only just been fixed, so if the very latest nightly crashes for you on the device when starting the RenderDoc app, try going back to `9e3a454a 2018-05-23`. The relevant fixes were made earlier this month or last month so trying that build should still work fine.

> `No library found in /data/app/net.osomi.Osomi_Core-2//lib/*/libVkLayer_GLES_RenderDoc.so for net.osomi.Osomi_Core - assuming injection is required.` seems weird to me, as I am not even using GLES, I am using Vulkan.

To clarify, it's impossible (to my knowledge) to tell in advance what API an app will use, so RenderDoc has to hook and inject to prepare to capture both.

> If I forcefully run my program and check my log function, it tells me it can't find a Vulkan layer. I think it is referring to libVkLayer_GLES_RenderDoc.so, as it mentions it in the RenderDoc log.
> 
> This keeps on happening if I open my app, even if RenderDoc is only on in the background, but when I close it, it just runs normally.

Yes this is unfortunately because the Android vulkan loader is severely limited, and it's not possible to launch Vulkan applications while RenderDoc is actively configured. As you found, if you want to launch your program you'd need to deactivate the RenderDoc connection so that it will remove the vulkan layer enabling command. This is an Android limitation so there's no realistic fix possible.

**Comment by Nielsbishere** [comment]
Yeah, the latest nightly works. Thank you!
